src/ConfigCatClientCache.ts

```diff
@@ -0,0 +1,52 @@
+import { IConfigCatKernel } from "./index";
+import { ConfigCatClientOptions } from "./ConfigCatClientOptions";
+import { ConfigCatClient } from "./ConfigCatClient";
+
+export class ConfigCatClientCache {
+    private instances: Record<string, [WeakRef<ConfigCatClient>, object]> = {};
+
+    public getOrCreate(options: ConfigCatClientOptions, configCatKernel: IConfigCatKernel): [ConfigCatClient, boolean] {
+        let instance: ConfigCatClient | undefined;
+
+        let cachedInstance = this.instances[options.apiKey];
+        if (cachedInstance) {
+            const [weakRef] = cachedInstance;
+            instance = weakRef.deref();
+            if (instance) {
+                return [instance, true];
+            }
+        }
+
+        const token = {};
+        instance = new ConfigCatClient(options, configCatKernel, token);
+        this.instances[options.apiKey] = cachedInstance = [new WeakRef(instance), token];
+        return [instance, false];
+    }
+
+    public remove(sdkKey: string, cacheToken: object) {
+        const cachedInstance = this.instances[sdkKey];
+
+        if (cachedInstance) {
+            const [weakRef, token] = cachedInstance;
+            const instanceIsAvailable = !!weakRef.deref();
```

Comment:
Why it is necessary to check `.deref()` function?

Reply:
You mean the return value of the `deref()` function, don't you? I check it to determine if the referenced object is still alive:

> The target object of the WeakRef, or undefined if the object has been garbage-collected.

(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef/deref#return_value)

Reply:
I mean the `!!` operator...I can not see why it is necessary. Based on your link, It should return an object or undefined at all times.

Reply:
`!!` is for converting the return value into a boolean. (Not only uses the function this information internally but it also needs to return it.)